src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+    decode(sp.update_time, bl);
+    decode(sp.owner, bl);
+    decode(sp.in_bl, bl);
+    DECODE_FINISH(bl);
+  }
+
+  //---------------------------------------------------------------------------
+  int init_dedup_pool_ioctx(RGWRados                 *rados,
+			    const DoutPrefixProvider *dpp,
+			    librados::IoCtx          &ioctx)
+  {
+    rgw_pool dedup_pool(DEDUP_POOL_NAME);
+    std::string pool_name(DEDUP_POOL_NAME);
+    // using Replica-1 for the intermediate data
+    // since it can be regenerated in case of a failure
+    std::string replica_count(std::to_string(1));
+    librados::bufferlist inbl;
+    std::string output;
+    std::string command = R"(
+    {
+      "prefix": "osd pool create",
+      "pool": ")" + pool_name +
+      R"(",
+      "pool_type": "replicated",
+      "size": )" + replica_count +
+      R"(
+    })";
+
+    int ret = rados->get_rados_handle()->mon_command(command, inbl, nullptr, &output);
+    if (output.length()) {
+      if (output != "pool 'rgw_dedup_pool' already exists") {
+	ldpp_dout(dpp, 10) << __func__ << "::" << output << dendl;
+      }
+    }
+    if (ret != 0 && ret != -EEXIST) {
+      ldpp_dout(dpp, 1) << __func__ << "::ERR: failed to create pool "
+			<< DEDUP_POOL_NAME << " with: "
+			<< cpp_strerror(ret) << ", ret=" << ret << dendl;
+      return ret;
+    }
+
+    ret = rgw_init_ioctx(dpp, rados->get_rados_handle(), dedup_pool, ioctx);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1) << "failed to initialize pool for listing with: "
+			<< cpp_strerror(ret) << dendl;
+    }
+
+    return ret;
+  }
+
+  //==========================================================================
+
+  //---------------------------------------------------------------------------
+  void cluster::clear()
+  {
+    d_curr_md5_shard = 0;
+    d_curr_worker_shard = 0;
+
+    d_num_completed_workers = 0;
+    d_num_completed_md5 = 0;
+
+    memset(d_completed_workers, TOKEN_STATE_PENDING, sizeof(d_completed_workers));
+    memset(d_completed_md5, TOKEN_STATE_PENDING, sizeof(d_completed_md5));
+
+    d_total_ingressed_obj = 0;
+    d_num_failed_workers = 0;
+  }
+
+
+  static constexpr auto COOKIE_LEN = 15;
+  static constexpr auto CLUSTER_ID_LEN = 15;
+  //---------------------------------------------------------------------------
+  cluster::cluster(const DoutPrefixProvider *_dpp,
+		   CephContext *cct,
+		   rgw::sal::Driver* driver):
+    dpp(_dpp),
+    d_lock_cookie(gen_rand_alphanumeric(cct, COOKIE_LEN)),
+    d_cluster_id (gen_rand_alphanumeric(cct, CLUSTER_ID_LEN))
+  {
+    clear();
+    ldpp_dout(dpp, 10) << __func__ << "::cluser_id=" << d_cluster_id << dendl;
+
+    auto store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!store) {
+      ldpp_dout(dpp, 0) << "ERR: failed dynamic_cast to RadosStore" << dendl;
+      // TBD: should we through an exception ??
+      return;
+    }
+
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
```

Comment:
these are real error cases. in the ctor you can either throw an excpetion, or plit that logic into an `init()` function with a result code, called after the ctor

Reply:
done